cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -357,16 +361,19 @@ public synchronized void stop(ProgramId programId) throws Exception {
    * Stops the specified run of the specified program.
    *
    * @param programId the {@link ProgramId program} to stop
-   * @param runId the runId of the program run to stop. If null, the first run of the program as returned by
-   *              {@link ProgramRuntimeService} is stopped.
+   * @param runId the runId of the program run to stop. If null, all runs of the program as returned by
+   *              {@link ProgramRuntimeService} are stopped.
    * @throws NotFoundException if the app, program or run was not found
    * @throws BadRequestException if an attempt is made to stop a program that is either not running or
    *                             was started by a workflow
    * @throws InterruptedException if there was a problem while waiting for the stop call to complete
    * @throws ExecutionException if there was a problem while waiting for the stop call to complete
    */
   public void stop(ProgramId programId, @Nullable String runId) throws Exception {
-    issueStop(programId, runId).get();
+    List<ListenableFuture<ProgramController>> issueStops = issueStop(programId, runId);
+    for (ListenableFuture<ProgramController> issueStop : issueStops) {
+      issueStop.get();
```

Comment:
This method is called from endpoints "/apps/{app-id}/{type}/{id}/runs/{run-id}/stop" or "/apps/{app-id}/{type}/{id}/{action}", it will only send back a HttpResponseStatus without any info. So if any exception happens, we should throw it and send a response back.